separate out dynamo .requires_grad and .is_grad_enabled guards

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101219
* __->__ #100570



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

bdhirsh committed:
[test, not ready] run inductor code under no_grad

[ghstack-poisoned]

bdhirsh committed:
Update on "[test, not ready] run inductor code under no_grad"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
The other reason why this shouldn't have happened in the first place, is because if we're executing inside a custom autograd function, grad should have been disabled anyway. Except, maybe not, because of reasons? (cc @albanD) I agree your approach should work but perhaps we should do this inside AOTAutograd, rather than in inductor.

bdhirsh committed:
Update on "[test, not ready] separate out dynamo .requires_grad and .is_grad_enabled guards"

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
Update: this is actually a guard bug. In the internal model repro that I have, we end up executing an inference graph when `torch.is_grad_enabled() == True` (this causes us to raise an autograd error when we try to run the out= operator). Instead, we should be failing a guard somewhere, and re-compiling a training graph. I still need to create a repro / test to accompany this PR though (haven't done that yet).

bdhirsh committed:
Update on "[test, not ready] separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "[test, not ready] separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -142,6 +149,7 @@ class TensorCheck {
   // necessarily capture device indices correctly.
   at::DeviceIndex device_index_;
   bool requires_grad_;
+  bool grad_mode_enabled_;
```
Do we actually need grad_mode_enabled_ here? It seems wholly redundant w GRAD_MODE (I know this is a prior issue with the code, and not something you introduced).

Would you be open to pulling it out?  the `GRAD_MODE` should cover all cases here.

bdhirsh replied:
yep happy to pull it out.

I see there's already a `GRAD_MODE` guard defined here. Does that mean it's enough just to yank this bool and only guard on `requires_grad` as part of the tensor guard? Or do you suspect there's something else we need. I can also yank it and see if my new tests still pass.

voznesenskym replied:
>  Does that mean it's enough just to yank this bool and only guard on requires_grad as part of the tensor guard? 

That should suffice. If not, we need to figure out why `GRAD_MODE` isn't working.

bdhirsh replied:
The tests fail when I remove it.

> If not, we need to figure out why GRAD_MODE isn't working.

It looks like the only use of the `GRAD_MODE` guard is in `GradModeVariable` [here](https://github.com/pytorch/pytorch/blob/96f46316c9d4788adbfdff531d0c35ecdcf316e9/torch/_dynamo/variables/ctx_manager.py#L114). That will probably be created if we see a `no_grad()` block **inside** the compiled region. But it looks like we aren't guarding on whatever the current grad state was when we first started compiling. Does that seem like the right thing to go fix (and any pointers?)

ezyang replied:
You can just unconditionally add a GRAD_MODE guard at the start of tracing in output_graph.py, similar to GuardBuilder.DETERMINISTIC_ALGORITHMS

bdhirsh replied:
thanks for the pointer! Updated, tests now pass

anijain2305 commented:
Btw, just want to point out that `moco` also has similar error signature

```
python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --inductor --device cuda --training --float32 --total-partitions 1 --partition-id 0 --log-graph-breaks --only=moco
```

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -32,7 +32,8 @@ class TensorCheck {
         device_index_(v.device().index()),
-        requires_grad_(state.grad_mode_enabled && v.requires_grad()),
+        requires_grad_(v.requires_grad()),
+        grad_mode_enabled_(state.grad_mode_enabled),
```
It's weird that we are going to record this on every single TensorCheck lol

bdhirsh replied:
yep that was Voz's feedback too - just updated, so we no longer stash grad mode state on the tensor guard, and instead we just have a single guard on the initial grad mode state when we start compiling.

anijain2305 commented:
This is the repro from `moco`. I think the problem might be more complicated (or this is just a separate bug that we have to solve). 

~~~
import torch

@torch.no_grad()
def gn(z):
    torch._dynamo.graph_break()
    return torch.sin(z)

def fn(x, y, z):
    z = gn(z)
    a = torch.mm(x, y)
    z = gn(z)
    return a

opt_fn = torch.compile(fn)

x = torch.randn(4, 4, requires_grad=True)
y = torch.randn(4, 4, requires_grad=True)
z = torch.randn(4)

opt_fn(x, y, z).sum().backward()

~~~

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
@anijain2305 Looks like this PR doesn't fix the repro you posted - definitely feels like the same vein though.

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_inference.csv:
```diff
@@ -25,7 +25,7 @@ hf_Bert_large,pass,0
 hf_GPT2,pass,0
 hf_Reformer,pass,27
-hf_T5_generate,pass,49
+hf_T5_generate,fail_to_run,48
```
rip lol

PaliC commented:
@pytorchbot revert -c "nosignal" -m "breaking inductor tests along with #101219"
Add ciflow/inductor to run the tests on the pull request. It seems like this PR is the one actually breaking inductor.

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang reviewed (approved):
stampy

bdhirsh commented:
Although I seem to be able to repro the IMA on the commit before my PR too 🤔 continuing to look.

ezyang commented:
I think nanogpt is just failing on main. There is some problem with the HF GPT checkpoint which they recently fixed, see https://github.com/huggingface/transformers/pull/23256

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "separate out dynamo .requires_grad and .is_grad_enabled guards"

Fixes https://github.com/pytorch/pytorch/issues/100977

This will hopefully fix this error (from [issue](https://github.com/pytorch/pytorch/issues/99616))

This PR fixes an internal model: we were running an inductor inference graph, but `torch.is_grad_enabled()` was True, causing us to error inside of the inference graph when we encountered an out= operator.

I haven't been able to create a smaller repro - before landing this, I want to create a smaller repro to convince myself of why we need to separate out these guards.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jon-chuang commented on torch/_dynamo/output_graph.py:
```diff
@@ -249,6 +250,8 @@ def __init__(
         )
 
+        self.guards.add(GradModeSource().make_guard(GuardBuilder.GRAD_MODE))
+
```
Out of curiosity, is the sole purpose of this guard for logging purposes?

The actual guard logic exists in the GlobalStateGuard which is opaque to how we register guard metadata, right?
